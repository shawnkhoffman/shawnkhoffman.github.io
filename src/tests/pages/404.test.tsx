import { render, screen, cleanup, within } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach } from 'vitest';
import userEvent from '@testing-library/user-event';
import { MemoryRouter } from 'react-router-dom';
import NotFound from '../../pages/404';
import '@testing-library/jest-dom';

const renderNotFound = () => {
    cleanup();
    return render(
        <MemoryRouter>
            <NotFound />
        </MemoryRouter>
    );
};

describe('NotFound Component', () => {
    beforeEach(() => {
        vi.resetAllMocks();
    });

    describe('Rendering', () => {
        it('matches snapshot', () => {
            vi.spyOn(Math, 'random').mockReturnValue(0);
            const { container } = renderNotFound();
            expect(container).toMatchSnapshot();
        });
    });

    describe('Interactions', () => {
        it('changes content when "Next Meme" button is clicked', async () => {
            const user = userEvent.setup();
            
            vi.spyOn(Math, 'random').mockReturnValueOnce(0).mockReturnValueOnce(1);
            
            const { container } = renderNotFound();
            
            const nextButton = screen.getByRole('button', { name: /Show next meme/i });
            const initialContent = container.innerHTML;
            
            await user.click(nextButton);
            
            expect(container.innerHTML).not.toBe(initialContent);
        });

        it('changes content when clicked multiple times', async () => {
            const user = userEvent.setup();
            
            vi.spyOn(Math, 'random')
                .mockReturnValueOnce(0)
                .mockReturnValueOnce(1)
                .mockReturnValueOnce(2);
            
            const { container } = renderNotFound();
            const nextButton = screen.getByRole('button', { name: /Show next meme/i });
            
            const initialContent = container.innerHTML;
            await user.click(nextButton);
            
            expect(container.innerHTML).not.toBe(initialContent);
        });
    });

    describe('Accessibility', () => {
        it('ensures interactive elements are focusable', () => {
            renderNotFound();
            const container = screen.getByTestId('404-container');

            const homeLink = within(container).getByRole('link', { name: /home/i });
            const nextButton = within(container).getByRole('button', { name: /next/i });

            expect(homeLink).toBeInTheDocument();
            expect(nextButton).toBeInTheDocument();
            expect(homeLink).not.toHaveAttribute('tabindex', '-1');
            expect(nextButton).not.toHaveAttribute('tabindex', '-1');
        });
    });

    describe('Structure', () => {
        it('has correct HTML structure', () => {
            renderNotFound();
            const container = screen.getByTestId('404-container');
            
            expect(container).toBeInTheDocument();
            expect(container.querySelector('img')).toBeInTheDocument();
            expect(container.querySelector('h1')).toBeInTheDocument();
            expect(container.querySelector('.flex')).toHaveClass('flex-col', 'sm:flex-row');
        });
    });
});
