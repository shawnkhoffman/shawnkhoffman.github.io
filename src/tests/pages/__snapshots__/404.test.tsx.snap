// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotFound Component > Structure > has correct HTML structure 1`] = `
<div>
  <div
    class="flex flex-col justify-center items-center text-center bg-base-200 p-4"
    data-testid="404-container"
  >
    <img
      alt="A funny gif representing a 404 error"
      class="w-full max-w-xs sm:max-w-md rounded-lg shadow-lg mb-4"
      src="https://i.giphy.com/media/v1.Y2lkPTc5MGI3NjExbGZmamRkbmkzM292OHdramZ4eHhhdjdvajN4NHMwdzRiMGtwMnFzNyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/bGgsc5mWoryfgKBx1u/giphy.gif"
    />
    <h1
      class="text-2xl sm:text-4xl font-bold mb-4"
    >
      404 - Page Not Found
    </h1>
    <p
      class="text-base sm:text-lg mb-6"
    >
      Looks like the code broke... again.
    </p>
    <div
      class="flex flex-col sm:flex-row gap-4 w-full justify-center"
    >
      <a
        aria-label="Return to the homepage"
        class="btn btn-primary w-full sm:w-auto"
        href="/"
      >
        Go back Home
      </a>
      <button
        aria-label="Show next meme"
        class="btn btn-secondary w-full sm:w-auto"
      >
        Next Meme
      </button>
    </div>
  </div>
</div>
`;
